src/rocketmq_producer.erl

```diff
@@ -85,50 +88,86 @@ batch_send_sync(Pid, Messages, Timeout) ->
 %% gen_server callback
 %%--------------------------------------------------------------------
 init([QueueId, Topic, Server, ProducerGroup, ProducerOpts]) ->
+    SSLOpts = maps:get(ssl, ProducerOpts, undefined),
     State = #state{producer_group = ProducerGroup,
                    topic = Topic,
                    queue_id = QueueId,
                    callback = maps:get(callback, ProducerOpts, undefined),
                    batch_size = maps:get(batch_size, ProducerOpts, 0),
                    server = Server,
                    opts = maps:get(tcp_opts, ProducerOpts, []),
+                   ssl_opts = SSLOpts,
+                   send_sock_mod = case SSLOpts of
+                                       undefined -> gen_tcp;
+                                       _ -> ssl
+                                   end,
                    producer_opts = ProducerOpts
                    },
     self() ! connecting,
     {ok, idle, State}.
 
-idle(_, connecting, State = #state{opts = Opts, server = Server}) ->
+idle(_, connecting, State = #state{opts = Opts, ssl_opts = SSLOpts, server = Server}) ->
     {Host, Port} = parse_url(Server),
     case gen_tcp:connect(Host, Port, merge_opts(Opts, ?TCPOPTIONS), ?TIMEOUT) of
         {ok, Sock} ->
             tune_buffer(Sock),
             gen_tcp:controlling_process(Sock, self()),
             start_keepalive(),
-            {next_state, connected, State#state{sock = Sock}};
+            ClientID = client_id(Sock),
+            case maybe_get_tls_sock(Sock, SSLOpts) of
+                {error, SSLError} ->
+                    {stop, {shutdown, SSLError}, State};
+                Sock1 ->
+                    {next_state, connected, State#state{sock = Sock1, client_id = ClientID}}
+            end;
         Error ->
             {stop, {shutdown, Error}, State}
     end;
 
 idle(_, ping, State = #state{sock = undefined}) ->
     {keep_state, State}.
 
+maybe_get_tls_sock(Sock, undefined) ->
+    Sock;
+maybe_get_tls_sock(Sock, SSLOpts) ->
+    case ssl:connect(Sock, SSLOpts, ?TIMEOUT) of
+        {ok, Sock1} ->
+            Sock1;
+        Error ->
+            Error
+    end.
+
 connected(_EventType, {tcp_closed, Sock}, State = #state{sock = Sock}) ->
     log_error("TcpClosed producer: ~p~n", [self()]),
     erlang:send_after(5000, self(), connecting),
     {next_state, idle, State#state{sock = undefined}};
 
+connected(_EventType, {ssl_closed, Sock}, State = #state{sock = Sock}) ->
+    log_error("SSLClosed producer: ~p~n", [self()]),
+    erlang:send_after(5000, self(), connecting),
+    {next_state, idle, State#state{sock = undefined}};
+
+connected(_EventType, {ssl_error, Sock, Reason}, State = #state{sock = Sock}) ->
+    log_error("SSL Socket Error, producer: ~p, reason: ~p~n", [self(), Reason]),
+    erlang:send_after(5000, self(), connecting),
```

Comment:
would suggest to close the socket for safety.

Reply:
Ok, this sounds like a good suggestion since I'm not sure if the socket is guaranteed to be closed when we get the error. Thanks!